src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
+                .map(Value::from)
+        }
+        CreateFunctionDataType::I64 | CreateFunctionDataType::BIGINT => {
+            get_arrow_value::<arrow::datatypes::Int64Type>(args, row_ix, col_ix)
+                .map(Value::from)
+        }
+        CreateFunctionDataType::F32
+        | CreateFunctionDataType::FLOAT
+        | CreateFunctionDataType::REAL => {
+            get_arrow_value::<arrow::datatypes::Float32Type>(args, row_ix, col_ix)
+                .map(|val| Value::from(val.to_bits()))
+        }
+        CreateFunctionDataType::F64 | CreateFunctionDataType::DOUBLE => {
+            get_arrow_value::<arrow::datatypes::Float64Type>(args, row_ix, col_ix)
+                .map(|val| Value::from(val.to_bits()))
+        }
+        CreateFunctionDataType::TEXT
+        | CreateFunctionDataType::CHAR
+        | CreateFunctionDataType::VARCHAR => match args
+            .get(col_ix)
+            .unwrap()
+            .as_any()
+            .downcast_ref::<arrow::array::StringArray>(
+        ) {
+            Some(arr) => Ok(Value::from(arr.value(row_ix))),
+            None => Err(DataFusionError::Internal(format!(
+                "Error casting column {:?} to string array",
+                col_ix
+            ))),
+        },
+        CreateFunctionDataType::BOOLEAN => match args
+            .get(col_ix)
+            .unwrap()
+            .as_any()
+            .downcast_ref::<arrow::array::BooleanArray>(
+        ) {
+            Some(arr) => Ok(Value::from(arr.value(row_ix))),
+            None => Err(DataFusionError::Internal(format!(
+                "Error casting column {:?} to boolean array",
+                col_ix
+            ))),
+        },
+        // serialize decimal as string
+        CreateFunctionDataType::DECIMAL {
+            precision: _,
+            scale: _,
+        } => get_arrow_value::<arrow::datatypes::Decimal128Type>(args, row_ix, col_ix)
+            .map(|val| Value::from(val.to_string())),
+        // dates are represented as i32 integer internally, serialize them as such.
+        CreateFunctionDataType::DATE => {
+            get_arrow_value::<arrow::datatypes::Date32Type>(args, row_ix, col_ix)
+                .map(|val| Value::from(val.to_string()))
+        }
+        // timestamps are represented as i64 integers internally, serialize them as such.
+        CreateFunctionDataType::TIMESTAMP => get_arrow_value::<
+            arrow::datatypes::TimestampNanosecondType,
+        >(args, row_ix, col_ix)
+        .map(|val| Value::from(val.to_string())),
+    }
+}
+
+fn decode_udf_result_primitive_array<T>(
+    encoded_results: &Vec<Value>,
+    decoder: &dyn Fn(&Value) -> Result<T::Native>,
+) -> Result<PrimitiveArray<T>>
+where
+    T: ArrowPrimitiveType,
+{
+    let mut decoded_results = Vec::with_capacity(encoded_results.len());
+    for i in encoded_results {
+        decoded_results.push(Some(decoder(i)?));
+    }
+    Ok(decoded_results.iter().collect::<PrimitiveArray<T>>())
```

Comment:
Not very keen on mut + for + push here. Also, is the `Some(...)` in the loop body doing anything? We never add a `None` to the vector anyway, so it's not like we'd be discarding items that failed to decode. Is it possible to rewrite this as:

```suggestion
    encoded_results.iter().map(decoder).collect()
```

? The `map()` will turn an iterator of `Value` into an iterator of `Result<T::Native>` and the `collect()` will turn this into a `Result<PrimitiveArray<T>>` (stopping the iteration if an error occurs, see https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect).

or, if it doesn't compile:

```suggestion
    encoded_results.iter().map(decoder).collect::<Result<PrimitiveArray<T>>>()
```

Reply:
I struggled with this for a while, the resulting code is definitely a compromise:
- `decoder()` can return an error, in which case there's no point in continuing a loop. I don't know of a way to terminate `iter().collect()` early.
- One option is to collect a default value for the type, but I couldn't get the Rust compiler to accept my attempt to get this working (and in retrospect it seems cleaner to exist in case of error instead of packing a vector full of defaults).
- The `Some()` is actually required by the call to `collect()`, not entirely sure why tbh.

Reply:
Looks a lot nicer this way!